src/LibraryManager/Providers/BaseProvider.cs

```diff
@@ -0,0 +1,297 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Web.LibraryManager.Contracts;
+using Microsoft.Web.LibraryManager.LibraryNaming;
+
+namespace Microsoft.Web.LibraryManager.Providers
+{
+    /// <summary>
+    /// Default implenentation for a provider, since most provider implementations are very similar.
+    /// </summary>
+    internal abstract class BaseProvider : IProvider
+    {
+        private readonly CacheService _cacheService;
+
+        public BaseProvider(IHostInteraction hostInteraction, CacheService cacheService)
+        {
+            HostInteraction = hostInteraction;
+            _cacheService = cacheService;
+        }
+
+        #region IProvider implementation
+
+        /// <inheritdoc />
+        public abstract string Id { get; }
+
+        /// <inheritdoc />
+        public virtual string NuGetPackageId => "Microsoft.Web.LibraryManager.Build";
+
+        /// <inheritdoc />
+        public abstract string LibraryIdHintText { get; }
+
+        /// <inheritdoc />
+        public IHostInteraction HostInteraction { get; }
+
+        /// <inheritdoc />
+        public virtual bool SupportsLibraryVersions => true;
+
+        /// <inheritdoc />
+        public abstract ILibraryCatalog GetCatalog();
```

Comment:
`GetCatalog()` is expected to be an expensive operation, typically requiring a network call.  It should not be a property.

Reply:
Although, looking at those implementations, getting the catalog is not the expensive part, it's doing anything with it.  This might be worth considering.

Reply:
Since this is a change to the contract API, punting on it for this PR.